### PR TITLE
fix(mongodb): improvements to schema inference

### PR DESCRIPTION
**Pull request by tychoish**
Was poking around this to write docs, and saw some room for improvement.

### CODE REVIEW

File: crates/datasources/src/mongodb/infer.rs [hunk 1]
```diff
@@ -121,13 +126,15 @@ fn bson_to_arrow_type(depth: usize, bson: &Bson) -> Result<DataType> {
...
+        }
         Bson::Int32(_) => DataType::Float64,
         Bson::Int64(_) => DataType::Float64,
-        Bson::Timestamp(_) => DataType::Timestamp(TimeUnit::Microsecond, None), // TODO: Nanosecond
-        Bson::Binary(_) => DataType::Binary,                                    // TODO: Subtype?
+        Bson::Timestamp(_) => return Err(MongoError::UnsupportedBsonType("OplogTimestamp")),
+        Bson::Binary(_) => DataType::Binary, // TODO: Subtype?
         Bson::ObjectId(_) => DataType::Utf8,
-        Bson::DateTime(_) => DataType::Timestamp(TimeUnit::Microsecond, None), // TODO: Nanosecond
+        Bson::DateTime(_) => DataType::Timestamp(TimeUnit::Millisecond, None),
```
Comment (tychoish) [inline crates/datasources/src/mongodb/infer.rs:137]: the type specifies millisecond resolution. Should keep it at this fidelity.

File: crates/datasources/src/mongodb/infer.rs [hunk 2]
```diff
@@ -5,13 +5,15 @@ use mongodb::bson::{doc, Bson, Document};
...
+const SAMPLE_PCT: f32 = 0.05;
 
-const MAX_SAMPLE_SIZE: usize = 30;
+const MAX_SAMPLE_SIZE: usize = 100;
 const MIN_SAMPLE_SIZE: usize = 10;
 
 /// Recursion limit for inferring the schema for nested documents.
-const RECURSION_LIMIT: usize = 5;
+//
+// The MongoDB kernel rejects nesting of greater than 100.
```
Comment (scsmithr) [inline crates/datasources/src/mongodb/infer.rs:15]:
```suggestion
///
/// The MongoDB kernel rejects nesting of greater than 100.
```

File: crates/datasources/src/mongodb/infer.rs [hunk 3]
```diff
@@ -121,13 +126,15 @@ fn bson_to_arrow_type(depth: usize, bson: &Bson) -> Result<DataType> {
         Bson::Null => DataType::Null,
         Bson::RegularExpression(_) => DataType::Utf8,
         Bson::JavaScriptCode(_) => DataType::Utf8,
-        Bson::JavaScriptCodeWithScope(_) => DataType::Utf8,
+        Bson::JavaScriptCodeWithScope(_) => {
+            return Err(MongoError::UnsupportedBsonType("CodeWithScope"))
+        }
```
[the remark this answers is not in the document]
Comment (scsmithr) [reply crates/datasources/src/mongodb/infer.rs:131]: What about `JavaScriptCode`? Is that a normal thing to stick in mongo?

File: crates/datasources/src/mongodb/infer.rs [hunk 4]
```diff
@@ -121,13 +126,15 @@ fn bson_to_arrow_type(depth: usize, bson: &Bson) -> Result<DataType> {
...
         Bson::Int32(_) => DataType::Float64,
         Bson::Int64(_) => DataType::Float64,
-        Bson::Timestamp(_) => DataType::Timestamp(TimeUnit::Microsecond, None), // TODO: Nanosecond
-        Bson::Binary(_) => DataType::Binary,                                    // TODO: Subtype?
+        Bson::Timestamp(_) => return Err(MongoError::UnsupportedBsonType("OplogTimestamp")),
```
[the remark this answers is not in the document]
Comment (scsmithr) [reply crates/datasources/src/mongodb/infer.rs:134]: Makes sense. Seems similar to sql server's timestamp then. I had a assumed this was date time during the initial impl.

**Review by scsmithr — APPROVED** [review]

File: crates/datasources/src/mongodb/infer.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply crates/datasources/src/mongodb/infer.rs:134]: common mistake!

you handled the real date type correctly, so there's that.

File: crates/datasources/src/mongodb/infer.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tychoish) [reply crates/datasources/src/mongodb/infer.rs:131]: not normal, but it is at least a string, so there's not a great harm in supporting it.

**Pull request merged**